Add OCI annotations based on metadata attached to wasm payload

f4z3r opened the pull request:
When pushing an OCI artifact, attach metadata such as source, url, description, version, and licenses to the container image based on the metadata attached to the wasm payload that was constructed in the build step.

## Feature or Problem

See #78 

## Related Issues

#78

## Release Information

Next.

## Consumer Impact

OCI artifacts now contain metadata when pushed to a registry.

## Testing

### Unit Test(s)

TODO

### Acceptance or Integration
<!---
Indicate any changes or additions to the acceptance or integration test suite 
--->

### Manual Verification
<!---
Mandatory. Indicate the steps that you took to verify that this pull request works 
--->

f4z3r committed:
feat(oci): add annotations based on metadata attached to wasm payload

When pushing an OCI artefact, attach metadata such as source, url,
description, version, and licenses to the container image based on the
metadata attached to the wasm payload that was constructed in the build
step.

Relates-to: #78
Signed-off-by: Jakob Beckmann <f4z3r-github@pm.me>

f4z3r commented:
@ricochet was that what you were thinking? I wanted to quickly get your feedback before I start implementing some tests, as I am not 100% sure how to tackle unit testing for scenarios where I need a build artifact. I would also add additional annotations based on the remaining meatadata available on wasm payloads.

PS: the code changes in the OCI push logic I 1-to-1 copied from PR #53 (thanks @brooksmtownsend 🤣 ).

ricochet commented:
Thank you for taking this! Your approach looks right to me. I'm little surprised the underlying crates are missing helpers like the oci client not having a creator with annotations and wasm-tools today doesn't have the canonical oci labels like a metadata2oci helper, but this is something we could upstream from wash later. I'm not asking for changes, what you have here looks correct to me.

Little heads up that `author` will change to authors in an upcoming wasm-tools release:
https://github.com/bytecodealliance/wasm-tools/pull/2334/files

f4z3r commented:
Sweet, thanks for the update. I will have a look to see if there is really no utility function to generate the labels.

Thanks for the heads-up regarding the renaming of `author`. Then I will have a look later at implementing some tests, adding the remaining labels, and potentially having a look at what upstreaming the changes would entail 👍🏽

f4z3r commented:
Alright, I added the remaining annotations that are in the intersection of [`Metadata` fields](https://docs.rs/wasm-metadata/latest/wasm_metadata/struct.Metadata.html) and [pre-defined keys in the opencontainer spec](https://specs.opencontainers.org/image-spec/annotations/#pre-defined-annotation-keys). I also added an annotation for the creation time of the OCI artifact.

I have two questions:

1. I am very unsure how to test this, as I would essentially need a build artifact to test this. Assuming that `wasm-tools` works (IMO we should not test that), there is little to no logic that is actually implemented in the annotation mapping. Do you see any sensible way of adding tests for this, or would you say it is not necessary?
2. The logic in `crates/wash/src/oci.rs` is a different story. There I copied Brooks' tests over. However, some tests like `test_standard_opencontainer_annotations` seem to test whether HashMaps work? Or am I missing something obvious here? I would have removed these personally.

ricochet commented:
> Alright, I added the remaining annotations that are in the intersection of [`Metadata` fields](https://docs.rs/wasm-metadata/latest/wasm_metadata/struct.Metadata.html) and [pre-defined keys in the opencontainer spec](https://specs.opencontainers.org/image-spec/annotations/#pre-defined-annotation-keys). I also added an annotation for the creation time of the OCI artifact.
> 
> I have two questions:
> 
> 1. I am very unsure how to test this, as I would essentially need a build artifact to test this. Assuming that `wasm-tools` works (IMO we should not test that), there is little to no logic that is actually implemented in the annotation mapping. Do you see any sensible way of adding tests for this, or would you say it is not necessary?
> 2. The logic in `crates/wash/src/oci.rs` is a different story. There I copied Brooks' tests over. However, some tests like `test_standard_opencontainer_annotations` seem to test whether HashMaps work? Or am I missing something obvious here? I would have removed these personally.

Your tests seem largely sufficient to me although I am very tempted to follow the Beyoncé rule for testing wasm-tools Metadata field stability, although easy to argue we're mostly safe via types. Upstreaming which OCI annotations relate to which metadata fields is ultimately the best longterm defense.

brooksmtownsend reviewed (approved):
Looks really nice!! Very simple.
